### PR TITLE
Load decoder.onnx only when necessary

### PR DESCRIPTION
**Pull request by klakhov**
### Checklist

- [ ] I submit my changes into the `develop` branch
- [ ] I have created a changelog fragment 
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

File: cvat-ui/plugins/sam/src/ts/index.tsx [hunk 1]
```diff
@@ -183,90 +214,119 @@ const samPlugin: SAMPlugin = {
...
+                            });
+                    });
                 },
             },
         },
     },
     data: {
+        initialized: false,
         core: null,
+        worker: new Worker(new URL('./inference.worker', import.meta.url)),
```
Comment (bsekachev) [inline cvat-ui/plugins/sam/src/ts/index.tsx:329]: Please, reduce number of embeddings and lowResMasks stored in RAM as it done for efficientvit

File: cvat-ui/plugins/sam/src/ts/index.tsx [hunk 2]
```diff
@@ -183,90 +214,119 @@ const samPlugin: SAMPlugin = {
...
+                                    width: null,
+                                    x,
+                                    y,
+                                }));
+
+                                const feeds = modelData({
+                                    clicks: composedClicks,
+                                    tensor: plugin.data.embeddings.get(key) as Tensor,
+                                    modelScale,
+                                    maskInput: null,
```
Comment (bsekachev) [inline cvat-ui/plugins/sam/src/ts/index.tsx:267]: maskInput should not always be null

File: cvat-ui/plugins/sam/src/ts/index.tsx [hunk 3]
```diff
@@ -183,90 +214,119 @@ const samPlugin: SAMPlugin = {
...
+                                }
+
+                                const modelScale = {
+                                    width: imWidth,
+                                    height: imHeight,
+                                    samScale: getModelScale(imWidth, imHeight),
+                                };
+
+                                const composedClicks = [...pos_points, ...neg_points].map(([x, y], index) => ({
+                                    clickType: index < pos_points.length ? 1 : 0 as 0 | 1 | -1,
```
Comment (bsekachev) [inline cvat-ui/plugins/sam/src/ts/index.tsx:255]:
```suggestion
                                    clickType: index < pos_points.length ? 1 : 0 as 0 | 1,
```

File: changelog.d/20240311_093929_klakhov_improve_sam.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+### Fixed
+
+- SAM decoder.onnx is loaded only when the model is used
```
Comment (bsekachev) [inline changelog.d/20240311_093929_klakhov_improve_sam.md:3]:
```suggestion
- Segment anything decoder is loaded anytime when CVAT is opened, but might be not required
```

File: cvat-ui/plugins/sam/src/ts/index.tsx [hunk 5]
```diff
@@ -154,27 +156,55 @@ const samPlugin: SAMPlugin = {
...
+                                        reject(new Error(
+                                            `Caught unexpected action response from worker: ${e.data.action}`,
+                                        ));
+                                    }
+
+                                    if (!e.data.error) {
+                                        samPlugin.data.initialized = true;
+                                        resolvePromise();
+                                    } else {
+                                        reject(new Error(`SAM worker was not initialized. ${e.data.error()}`));
```
Comment (bsekachev) [inline cvat-ui/plugins/sam/src/ts/index.tsx:191]: I do not think `e.data.error()` may be a function here

File: cvat-ui/plugins/sam/src/ts/index.tsx [hunk 6]
```diff
@@ -183,106 +213,135 @@ const samPlugin: SAMPlugin = {
...
+                                    width: null,
+                                    x,
+                                    y,
+                                }));
+
+                                const feeds = modelData({
+                                    clicks: composedClicks,
+                                    tensor: plugin.data.embeddings.get(key) as Tensor,
+                                    modelScale,
+                                    maskInput: plugin.data.lowResMasks.has(key) ?
```
Comment (bsekachev) [inline cvat-ui/plugins/sam/src/ts/index.tsx:266]: Also not the best solution, please look at efficientvit

Comment (bsekachev) [reply cvat-ui/plugins/sam/src/ts/index.tsx:266]: It is buggy when removing and optionally adding new points.

Comment (klakhov) [reply cvat-ui/plugins/sam/src/ts/index.tsx]: Added logic with `lastClicks`

**Review by bsekachev — APPROVED** [review]

**Pull request merged**